File: src/mathdoc/depgraph/algorithms.py
from collections.abc import Mapping
from collections.abc import Sequence


def find_cycle(
    dep_graph: Mapping[str, Sequence[str]],
    *,
    root_fnode: str | None = None,
) -> list[str] | None:
    state: dict[str, int] = {}
    stack: list[str] = []
    stack_idx: dict[str, int] = {}

    def dfs(fnode: str) -> list[str] | None:
        state[fnode] = 1
        stack_idx[fnode] = len(stack)
        stack.append(fnode)

        for dep_fnode in dep_graph.get(fnode, []):
            dep_state = state.get(dep_fnode, 0)
            if dep_state == 0:
                cycle = dfs(dep_fnode)
                if cycle is not None:
                    return cycle
            elif dep_state == 1:
                start = stack_idx[dep_fnode]
                return stack[start:] + [dep_fnode]

        stack.pop()
        stack_idx.pop(fnode, None)
        state[fnode] = 2
        return None

    roots = [root_fnode] if root_fnode is not None else list(dep_graph)
    for fnode in roots:
        if fnode not in dep_graph:
            continue
        if state.get(fnode, 0) != 0:
            continue
        cycle = dfs(fnode)
        if cycle is not None:
            return cycle
    return None


def topo_dependencies_first(
    dep_graph: Mapping[str, Sequence[str]],
    *,
    root_fnode: str,
) -> list[str]:
    visited: set[str] = set()
    order: list[str] = []

    def dfs(fnode: str) -> None:
        if fnode in visited:
            return
        visited.add(fnode)
        for dep_fnode in dep_graph.get(fnode, []):
            dfs(dep_fnode)
        order.append(fnode)

    dfs(root_fnode)
    return order


def strongly_connected_components(
    dep_graph: Mapping[str, Sequence[str]],
) -> list[list[str]]:
    index = 0
    stack: list[str] = []
    on_stack: set[str] = set()
    indices: dict[str, int] = {}
    lowlinks: dict[str, int] = {}
    components: list[list[str]] = []

    def strongconnect(fnode: str) -> None:
        nonlocal index
        indices[fnode] = index
        lowlinks[fnode] = index
        index += 1
        stack.append(fnode)
        on_stack.add(fnode)

        for dep_fnode in dep_graph.get(fnode, []):
            if dep_fnode not in indices:
                strongconnect(dep_fnode)
                lowlinks[fnode] = min(lowlinks[fnode], lowlinks[dep_fnode])
            elif dep_fnode in on_stack:
                lowlinks[fnode] = min(lowlinks[fnode], indices[dep_fnode])

        if lowlinks[fnode] != indices[fnode]:
            return

        component: list[str] = []
        while stack:
            member = stack.pop()
            on_stack.discard(member)
            component.append(member)
            if member == fnode:
                break
        components.append(component)

    for fnode in list(dep_graph):
        if fnode not in indices:
            strongconnect(fnode)
    return components


def component_has_cycle(
    dep_graph: Mapping[str, Sequence[str]],
    component: list[str],
) -> bool:
    if len(component) > 1:
        return True
    if not component:
        return False
    fnode = component[0]
    return fnode in dep_graph.get(fnode, [])


def representative_cycle(
    dep_graph: Mapping[str, Sequence[str]],
    component: list[str],
) -> list[str] | None:
    if not component:
        return None
    if len(component) == 1:
        fnode = component[0]
        if fnode in dep_graph.get(fnode, []):
            return [fnode, fnode]
        return None

    component_set = set(component)
    visited: set[str] = set()
    stack: list[str] = []
    stack_index: dict[str, int] = {}

    def dfs(fnode: str) -> list[str] | None:
        visited.add(fnode)
        stack_index[fnode] = len(stack)
        stack.append(fnode)

        for dep_fnode in dep_graph.get(fnode, []):
            if dep_fnode not in component_set:
                continue
            if dep_fnode not in visited:
                cycle = dfs(dep_fnode)
                if cycle is not None:
                    return cycle
            elif dep_fnode in stack_index:
                start = stack_index[dep_fnode]
                return stack[start:] + [dep_fnode]

        stack.pop()
        stack_index.pop(fnode, None)
        return None

    for fnode in sorted(component):
        if fnode in visited:
            continue
        cycle = dfs(fnode)
        if cycle is not None:
            return cycle
    return None

File: src/core/algorithms.rs
use std::collections::{HashMap, HashSet, VecDeque};

/// Compute every node's height from a pre-loaded dependency graph: leaves are 0,
/// and every acyclic parent is one greater than its deepest dependency.
/// Nodes in cycles retain the height accumulated before the cycle closes.
pub fn all_topo_depths(graph: &HashMap<String, Vec<String>>) -> HashMap<String, u32> {
    if graph.is_empty() {
        return HashMap::new();
    }

    let mut reverse: HashMap<&str, Vec<&str>> =
        graph.keys().map(|fnode| (fnode.as_str(), vec![])).collect();
    for (source, targets) in graph {
        for target in targets {
            if graph.contains_key(target.as_str()) {
                reverse
                    .entry(target.as_str())
                    .or_default()
                    .push(source.as_str());
            }
        }
    }

    let mut remaining: HashMap<&str, usize> = graph
        .iter()
        .map(|(fnode, targets)| {
            let count = targets
                .iter()
                .filter(|target| graph.contains_key(target.as_str()))
                .count();
            (fnode.as_str(), count)
        })
        .collect();
    let mut depths: HashMap<&str, u32> = graph.keys().map(|fnode| (fnode.as_str(), 0)).collect();
    let mut queue: std::collections::VecDeque<&str> = remaining
        .iter()
        .filter(|(_, count)| **count == 0)
        .map(|(fnode, _)| *fnode)
        .collect();

    while let Some(fnode) = queue.pop_front() {
        let depth = depths[fnode];
        for parent in reverse.get(fnode).into_iter().flatten() {
            if let Some(parent_depth) = depths.get_mut(parent) {
                *parent_depth = (*parent_depth).max(depth + 1);
            }
            let count = remaining.entry(parent).or_default();
            *count = count.saturating_sub(1);
            if *count == 0 {
                queue.push_back(parent);
            }
        }
    }

    depths
        .into_iter()
        .map(|(fnode, depth)| (fnode.to_string(), depth))
        .collect()
}

/// Compute each selected node's weakly connected-component size.
///
/// Edges are treated as undirected, and graph nodes outside `members` are ignored.
pub fn weak_component_sizes(
    graph: &HashMap<String, Vec<String>>,
    members: &HashSet<String>,
) -> HashMap<String, u32> {
    let mut adjacency: HashMap<&str, HashSet<&str>> = members
        .iter()
        .map(|member| (member.as_str(), HashSet::new()))
        .collect();
    for (source, targets) in graph {
        if !members.contains(source) {
            continue;
        }
        for target in targets {
            if !members.contains(target) {
                continue;
            }
            adjacency
                .entry(source.as_str())
                .or_default()
                .insert(target.as_str());
            adjacency
                .entry(target.as_str())
                .or_default()
                .insert(source.as_str());
        }
    }

    let mut sizes = HashMap::new();
    let mut seen = HashSet::new();
    for start in members {
        if seen.contains(start.as_str()) {
            continue;
        }
        let mut component = Vec::new();
        let mut queue = VecDeque::from([start.as_str()]);
        while let Some(node) = queue.pop_front() {
            if !seen.insert(node) {
                continue;
            }
            component.push(node);
            for neighbor in adjacency.get(node).into_iter().flatten() {
                if !seen.contains(neighbor) {
                    queue.push_back(neighbor);
                }
            }
        }

        let size = component.len() as u32;
        sizes.extend(component.into_iter().map(|node| (node.to_string(), size)));
    }
    sizes
}

/// Compute all strongly connected components using Kosaraju's algorithm.
/// Returns a list of components; each component is a list of fnodes.
pub fn strongly_connected_components(dep_graph: &HashMap<String, Vec<String>>) -> Vec<Vec<String>> {
    strongly_connected_component_refs(dep_graph)
        .into_iter()
        .map(|component| component.into_iter().map(str::to_string).collect())
        .collect()
}

fn strongly_connected_component_refs(dep_graph: &HashMap<String, Vec<String>>) -> Vec<Vec<&str>> {
    // Step 1: DFS on original graph; collect finish order
    let mut visited: HashSet<&str> = HashSet::new();
    let mut finish_order: Vec<&str> = Vec::new();

    for start in dep_graph.keys() {
        if visited.contains(start.as_str()) {
            continue;
        }
        let mut stack: Vec<(&str, bool)> = vec![(start.as_str(), false)];
        while let Some((node, done)) = stack.pop() {
            if done {
                finish_order.push(node);
            } else if !visited.contains(node) {
                visited.insert(node);
                stack.push((node, true));
                let children = dep_graph.get(node).map(Vec::as_slice).unwrap_or_default();
                for child in children.iter().rev() {
                    if !visited.contains(child.as_str()) {
                        stack.push((child.as_str(), false));
                    }
                }
            }
        }
    }

    // Step 2: Build transpose graph
    let mut transpose: HashMap<&str, Vec<&str>> = HashMap::new();
    for (src, dsts) in dep_graph {
        for dst in dsts {
            transpose
                .entry(dst.as_str())
                .or_default()
                .push(src.as_str());
        }
    }

    // Step 3: DFS on transpose in reverse finish order; each tree = one SCC
    let mut visited2: HashSet<&str> = HashSet::new();
    let mut components: Vec<Vec<&str>> = Vec::new();

    for &start in finish_order.iter().rev() {
        if visited2.contains(start) {
            continue;
        }
        let mut component = Vec::new();
        let mut stack: Vec<&str> = vec![start];
        while let Some(node) = stack.pop() {
            if visited2.contains(node) {
                continue;
            }
            visited2.insert(node);
            component.push(node);
            let children = transpose.get(node).map(Vec::as_slice).unwrap_or_default();
            for &child in children.iter().rev() {
                if !visited2.contains(child) {
                    stack.push(child);
                }
            }
        }
        components.push(component);
    }

    components
}

/// Return one representative cycle from each cyclic strongly connected component.
pub fn representative_cycles(dep_graph: &HashMap<String, Vec<String>>) -> Vec<Vec<String>> {
    let mut cycles: Vec<Vec<String>> = strongly_connected_component_refs(dep_graph)
        .into_iter()
        .filter_map(|component| representative_cycle(dep_graph, &component))
        .collect();
    cycles.sort();
    cycles
}

/// Find a representative cycle within a strongly connected component.
/// Returns the cycle as a list of fnodes (first == last), or None if no cycle.
fn representative_cycle(
    dep_graph: &HashMap<String, Vec<String>>,
    component: &[&str],
) -> Option<Vec<String>> {
    if component.is_empty() {
        return None;
    }
    if component.len() == 1 {
        let fnode = component[0];
        if dep_graph
            .get(fnode)
            .map(|dependencies| dependencies.iter().any(|dependency| dependency == fnode))
            .unwrap_or(false)
        {
            return Some(vec![fnode.to_string(), fnode.to_string()]);
        }
        return None;
    }

    let component_set: HashSet<&str> = component.iter().copied().collect();
    let mut visited: HashSet<&str> = HashSet::new();
    let mut path: Vec<&str> = Vec::new();
    let mut path_idx: HashMap<&str, usize> = HashMap::new();

    // Iterate in sorted order to match Python spec behavior
    let mut sorted_component = component.to_vec();
    sorted_component.sort();

    for &start in &sorted_component {
        if visited.contains(start) {
            continue;
        }

        let mut dfs_stack: Vec<(&str, usize)> = vec![(start, 0)];
        visited.insert(start);
        path_idx.insert(start, path.len());
        path.push(start);

        while let Some(frame) = dfs_stack.last_mut() {
            let fnode = frame.0;
            let children = dep_graph.get(fnode).map(Vec::as_slice).unwrap_or_default();

            // Find next child within the component
            let mut found = false;
            while frame.1 < children.len() {
                let child = children[frame.1].as_str();
                frame.1 += 1;
                if !component_set.contains(child) {
                    continue;
                }
                if !visited.contains(child) {
                    visited.insert(child);
                    path_idx.insert(child, path.len());
                    path.push(child);
                    dfs_stack.push((child, 0));
                    found = true;
                    break;
                } else if path_idx.contains_key(child) {
                    // Back edge within current path: cycle found
                    let start_idx = path_idx[child];
                    let mut cycle = path[start_idx..]
                        .iter()
                        .map(|node| node.to_string())
                        .collect::<Vec<_>>();
                    cycle.push(child.to_string());
                    return Some(cycle);
                }
            }

            if !found {
                dfs_stack.pop();
                path.pop();
                path_idx.remove(&fnode);
            }
        }
    }
    None
}
